### PR TITLE
Support faiss FLAT index get vector with metric COSINE

### PR DESCRIPTION
**Pull request by cydrain**
Issue: #79

### CODE REVIEW

File: thirdparty/faiss/faiss/IndexFlat.cpp [hunk 1]
```diff
@@ -36,7 +38,11 @@ void IndexFlat::search(
 
     if (metric_type == METRIC_INNER_PRODUCT) {
         float_minheap_array_t res = {size_t(n), size_t(k), labels, distances};
-        knn_inner_product(x, get_xb(), d, n, ntotal, &res, bitset);
+        if (is_cosine) {
+            knn_cosine(x, get_xb(), d, n, ntotal, &res, bitset);
```
Comment (hhy3) [inline thirdparty/faiss/faiss/IndexFlat.cpp:42]: knn_cosine is a slow function which is a compromise for brute force, and will compute the norm each time distance is calculated. It's better to pre-compute norms of all base vectors and use them

Comment (cydrain) [reply thirdparty/faiss/faiss/IndexFlat.cpp:42]: I know that, this is a middle-solution, I'll roll back some un-necessary faiss code change, then implement the flat with norm data solution

File: src/index/flat/flat.cc [hunk 2]
```diff
@@ -236,7 +237,7 @@ class FlatIndexNode : public IndexNode {
     bool
     HasRawData(const std::string& metric_type) const override {
         if constexpr (std::is_same<T, faiss::IndexFlat>::value) {
-            return !IsMetricType(metric_type, metric::COSINE);
+            return true;
         }
         if constexpr (std::is_same<T, faiss::IndexBinaryFlat>::value) {
             return true;
```
Comment (zhengbuqian) [inline src/index/flat/flat.cc:243]: curious: why doesn't this if have an else branch? this doesn't seem legit c++ code to me.. how come it compiles before?

**Review by zhengbuqian — COMMENTED** [review]
/approve

File: src/index/flat/flat.cc (hunk 2 above)
Comment (cydrain) [reply src/index/flat/flat.cc:243]: There are only 2 possibilities to run into here, one is faiss::IndexFlat, the other is faiss::IndexBinaryFlat, so no else

**Review by hhy3 — COMMENTED** [review]
/lgtm

File: thirdparty/faiss/faiss/IndexFlat.cpp [hunk 3]
```diff
@@ -18,10 +18,28 @@
...
-        : IndexFlatCodes(sizeof(float) * d, d, metric) {}
+IndexFlat::IndexFlat(idx_t d, MetricType metric, bool is_cosine)
+        : IndexFlatCodes(sizeof(float) * d, d, metric) {
+    this->is_cosine = is_cosine;
+}
+
+void IndexFlat::add(idx_t n, const float* x) {
+    FAISS_THROW_IF_NOT(is_trained);
+    codes.resize((ntotal + n) * code_size);
+    sa_encode(n, x, &codes[ntotal * code_size]);
```
Comment (liliu-z) [inline thirdparty/faiss/faiss/IndexFlat.cpp:33]: Why don't call `Index::add()` here and handle norm after this calling

**Comment by liliu-z** [comment]
issue: 26960

**Pull request merged**